File: web/src/lib/schemas.ts
/**
 * TypeScript types and Zod schemas for movies
 * Matches backend Pydantic schemas
 */

import { z } from "zod";

// Zod schema for movie creation (matches backend)
export const movieCreateSchema = z.object({
  name: z.string().min(1, "Name is required").max(255),
  year: z.number().int().min(1800).max(2100),
  age_limit: z.number().int().min(0).max(18),
  rating: z.number().int().min(1).max(5),
  synopsis: z.string().optional(),
  genres: z.array(z.string()).min(1, "At least one genre is required"),
});

export type MovieCreate = z.infer<typeof movieCreateSchema>;

// TypeScript interfaces for responses
export interface Movie {
  id: number;
  name: string;
  year: number;
  age_limit: number;
  rating: number;
  synopsis: string | null;
  genres: string[];
}

export interface MoviesListResponse {
  total: number;
  items: Movie[];
}

export interface MovieFilters {
  search?: string;
  sort?: "year" | "rating" | "name";
  order?: "asc" | "desc";
  limit?: number;
  offset?: number;
}

File: api/app/schemas.py
"""Pydantic schemas for request/response validation."""

from pydantic import BaseModel, Field
from typing import List, Optional


class GenreBase(BaseModel):
    """Base schema for Genre."""
    name: str = Field(..., min_length=1, max_length=50)


class GenreResponse(GenreBase):
    """Genre response schema."""
    id: int

    class Config:
        from_attributes = True


class MovieBase(BaseModel):
    """Base schema for Movie with shared fields."""
    name: str = Field(..., min_length=1, max_length=255)
    year: int = Field(..., ge=1800, le=2100)
    age_limit: int = Field(..., ge=0, le=18)
    rating: int = Field(..., ge=1, le=5)
    synopsis: Optional[str] = None
    genres: List[str] = Field(..., min_length=1)


class MovieCreate(MovieBase):
    """Schema for creating a new movie."""
    pass


class MovieResponse(BaseModel):
    """Schema for movie responses."""
    id: int
    name: str
    year: int
    age_limit: int
    rating: int
    synopsis: Optional[str]
    genres: List[str]

    class Config:
        from_attributes = True

    @classmethod
    def from_orm_movie(cls, movie):
        """Convert SQLAlchemy Movie model to response schema."""
        return cls(
            id=movie.id,
            name=movie.name,
            year=movie.year,
            age_limit=movie.age_limit,
            rating=movie.rating,
            synopsis=movie.synopsis,
            genres=[genre.name for genre in movie.genres],
        )


class MoviesListResponse(BaseModel):
    """Paginated response for movie list."""
    total: int
    items: List[MovieResponse]
